pkgs/shells/fish/default.nix

```diff
@@ -35,28 +35,56 @@ let
     #     source both, but source the more global configuration files earlier
     #     than the more local ones, so that more local configurations inherit
     #     from but override the more global locations.
-    
+
     if test -f /etc/fish/config.fish
       source /etc/fish/config.fish
     end
-    
+
     #                                                                             #
     ############### ↑ Nix hook for sourcing /etc/fish/config.fish ↑ ###############
   '';
 
+
+  nixEnvironmentSetupScriptPrefix = "@NIX_STORE@/../var/nix/profiles/default/etc/profile.d";
```

Comment:
I’m not 100% sure `../var/nix` does always hold here.

Reply:
Any ideas on when it wouldn't? (I would love a more robust way of doing this if anyone can think of one.)

Reply:
Even more interesting: there is no `/default/etc/profile.d` on my system. Where/When should it exist? Can we factor that logic out to be more semantic instead of doing brittle bash if/then/else checks?

Reply:
Well, NixOS isn't a distinct platform in Nixpkgs, right? There's the additional problem that on macOS, either can appear. With a plain Nix install on macOS, we'll have the `profile.d/nix.sh` script, but if one is using nix-darwin, they'll also have a module system like on NixOS. If on that platform we always read the `nix.sh` file, things controlled in `setEnvironment` (like the environment variables relating to whether and where to find the Nix daemon) will be ignored.

Reply:
@Profpatsch I was just looking at the Nix package in Nixpkgs and it seems like maybe we could add `stateDir` to `nix.passthru` and then use that in the Fish package. Does that make sense? Could we maybe expose it in the `stdenv` in its own right along side `$NIX_STORE` as something like `$NIX_STATE`?

Reply:
> maybe we could add stateDir to nix.passthru

I’m not sure we even need to find the state dir. Can you point me to documentation by what these shell files are generated on each system?
It feels to me like those are (undocumented) nix interna that shouldn’t be depended upon. So before we use them, the interface should be documented and standardized.

Reply:
> I’m not sure we even need to find the state dir. Can you point me to documentation by what these shell files are generated on each system?

The reason we need to find the state directory is because we need to find a Nix profile (the default profile, which gets the `nix.sh` and `nix-daemon.sh` files) before NIX_PROFILES is set. The state directory we only really need to find that default profile (instead of assuming it lives under `/nix/var/nix/...`, as that can change at compile time.

I don't think there's any documentation about it, but the files get put in the appropriate directories [here](https://github.com/NixOS/nix/blob/master/scripts/local.mk) and the templates for them are [here (for no Nix daemon)](https://github.com/NixOS/nix/blob/master/scripts/nix-profile.sh.in) and [here (to a Nix daemon)](https://github.com/NixOS/nix/blob/master/scripts/nix-profile-daemon.sh.in).
All I really need is a reliable place in which to find these shell scripts that Nix uses to set up the environment, both for NixOS and non-NixOS. I'm not invested in how it's exposed-- what I did here was find something that works for a typical Nix installation (i.e., any Nix which hasn't been compiled to use a non-default state directory).

> It feels to me like those are (undocumented) nix interna that shouldn’t be depended upon. So before we use them, the interface should be documented and standardized.

This is a fair point, and I should have followed through with it when I first read it. Who should I contact about this and how?